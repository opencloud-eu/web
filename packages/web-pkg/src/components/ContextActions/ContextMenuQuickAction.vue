<template>
  <oc-button
    :id="`context-menu-trigger-${resourceDomSelector(item)}`"
    v-oc-tooltip="contextMenuLabel"
    :aria-label="contextMenuLabel"
    appearance="raw"
    class="quick-action-button oc-ml-xs oc-p-xs"
    @click.stop.prevent="
      $emit('quickActionClicked', {
        event: $event,
        dropdown: $refs[`context-menu-drop-ref-${resourceDomSelector(item)}`]
      })
    "
  >
    <oc-icon name="more-2" />
    <oc-drop
      :ref="`context-menu-drop-ref-${resourceDomSelector(item)}`"
      :drop-id="`context-menu-drop-${resourceDomSelector(item)}`"
      :toggle="`#context-menu-trigger-${resourceDomSelector(item)}`"
      position="bottom-end"
      mode="click"
      padding-size="small"
      close-on-click
    >
      <slot name="contextMenu" :item="item" />
    </oc-drop>
  </oc-button>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import { extractDomSelector, Resource } from '@opencloud-eu/web-client'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'ContextMenuQuickAction',
  props: {
    item: {
      type: Object,
      required: true
    },
    resourceDomSelector: {
      type: Function,
      required: false,
      default: (resource: Resource) => extractDomSelector(resource.id)
    }
  },
  emits: ['quickActionClicked'],
  setup() {
    const { $gettext } = useGettext()
    const contextMenuLabel = computed(() => $gettext('Show context menu'))

    return {
      contextMenuLabel
    }
  }
})
</script>
