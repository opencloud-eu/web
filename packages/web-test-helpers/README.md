# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
