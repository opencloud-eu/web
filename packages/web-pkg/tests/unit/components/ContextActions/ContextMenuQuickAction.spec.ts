import ContextMenuQuickAction from '../../../../src/components/ContextActions/ContextMenuQuickAction.vue'
import { defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'

describe('ContextMenuQuickAction component', () => {
  it('renders component', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('triggers the "quickActionClicked"-event on click', async () => {
    const { wrapper } = getWrapper()
    await wrapper.find('button').trigger('click')
    expect(wrapper.emitted('quickActionClicked')).toBeTruthy()
  })
})

function getWrapper({ item = { id: '1' } } = {}) {
  return {
    wrapper: mount(ContextMenuQuickAction, {
      props: {
        item
      },
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }
}
