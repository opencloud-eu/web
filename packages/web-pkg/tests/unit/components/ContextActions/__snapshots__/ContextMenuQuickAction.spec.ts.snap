// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContextMenuQuickAction component > renders component 1`] = `
"<button type="button" aria-label="Show context menu" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw quick-action-button oc-ml-xs oc-p-xs" id="context-menu-trigger-1">
  <!--v-if-->
  <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m"><!----></span>
  <div id="context-menu-drop-1" class="oc-drop oc-box-shadow-medium oc-rounded">
    <div class="oc-card oc-card-body oc-p-s"></div>
  </div>
</button>"
`;
